src/Linux/local_cache.cpp

```diff
@@ -180,35 +183,48 @@ static void make_dir(const std::string& dirname, mode_t mode)
     }
 }
 
+static void load_cache_locations()
+{
+    cache_locations[0] = ::getenv("AZDCAP_CACHE");
+    cache_locations[1] = ::getenv("XDG_CACHE_HOME");
+    cache_locations[2] = ::getenv("HOME");
+    cache_locations[3] = ::getenv("TMPDIR");
+
+    // The fallback location isn't an environment variable
+    cache_locations[4] = "/tmp/";
+}
+
 static void init_callback()
 {
-    const char * env_home = ::getenv("HOME");
-    const char * env_azdcap_cache = ::getenv("AZDCAP_CACHE");
+    load_cache_locations();
     const std::string application_name("/.az-dcap-client/");
-    
     std::string dirname;
+    std::string all_locations;
 
-    if (env_azdcap_cache != 0 && (strcmp(env_azdcap_cache,"") != 0))
-    {
-        dirname = env_azdcap_cache;
-    }
-    else if (env_home != 0 && (strcmp(env_home,"") != 0))
+    // Try the cache locations in order
+    for (auto &cache_location : cache_locations)
     {
-        dirname = std::string(env_home);
+        if (cache_location != 0 && strcmp(cache_location, "") != 0)

```

Comment:
Maybe I'm misinterpreting the code, but this still seems to assume that the cache can always write to the cache location. I thought that the code was supposed to tolerate cache creation failures.

Reply:
The tolerance comes from the caller swallowing the runtime_error thrown by local_cache - I don't like to do it that way but it seemed to be the least intrusive